test/src/test/java/hudson/model/JobQueueTest.java

```diff
@@ -42,6 +43,63 @@ public void setUp() {
         RunListener.all().add(listener);
     }
 
+    /**
+    * Verify that queue items are counted correctly for labels even when they are not == comparable
+    */
+    @Test
+    public void projectLabelUsesEquals() throws Exception {
+        Queue queue = j.getInstance().getQueue();
+
+        final String LABELA = "LABELA";
+        final String LABELB = "LABELB";
+
+        final Label labelExpression = Label.get(LABELA + "&&" + LABELB);
+        final Label sameLabelExpression = Label.get(LABELA).and(Label.get(LABELB));
+
+        //These two conditions are not what is being tested, but are an assumption made during
+        //the rest of the test. If the two labels are == then the test is moot.
+        assertTrue("Expressions are equal", labelExpression.equals(sameLabelExpression));
+        assertTrue("Expressions are !=", labelExpression != sameLabelExpression);
+
+        //Use a special Project to test with specific label instance
+        LabelTestProject project = j.getInstance().createProject(LabelTestProject.class,"labelproject");
+
+        project.setAssignedLabel( sameLabelExpression );
+
+        //Kick the first Build
+        project.scheduleBuild2(1);
+
+        //The project should be in Queue when Run is in BUILDING stage
+        assertTrue(project.isInQueue());
+
+        // ensure the queued items are in the right state, i.e. buildable rather
+        // than waiting
+        {
+            int count = 0;
+            while (queue.getItem(project) instanceof WaitingItem && ++count < 100) {
+                queue.maintain();
+                Thread.sleep(10);
```

Comment:
I've looked into queue.maintain(), nothing jumps out at me as invoking other threads. Though it may be other threads modify the queue.

Do you think it is worth trying just increasing the time waiting? The only other option I can think of is learning the entire inner workings of the Queue and rewriting it to be more testable.